src/algebra/category/Module/change_of_rings.lean

```diff
@@ -0,0 +1,84 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import algebra.category.Module.basic
+import algebra.category.Algebra.basic
+import ring_theory.tensor_product
+
+/-!
+# Change Of Rings
+
+## Main definitions
+
+* `category_theory.Module.restrict_scalars`: given rings `R, S` and a ring homomorphism `R ⟶ S`,
+  then `restrict_scalars : Module S ⥤ Module R` is defined by `M ↦ M` where `M : S-module` is seen
+  as `R-module` by `r • m := f r • m` and `S`-linear map `l : M ⟶ M'` is `R`-linear as well.
+
+## List of notations
+Let `R, S` be rings and `f : R →+* S`
+* if `M` is an `S`-module, `r : R` and `m : M` then notation `r r•[f] m` means `R`-scalar action on
+  `M` defined by `f r • m`.
+-/
+
+
+namespace category_theory.Module
+
+universes v u₁ u₂
+
+namespace restrict_scalars
+
+variables {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)
+
+section unbundled
+
+variables (M : Type v) [add_comm_monoid M] [module S M]
+
+/-- The `R`-scalar multiplication on `S`-module M defined by `r • m := f r • m` -/
+protected def has_smul : has_smul R M :=
+(module.comp_hom M f).to_has_smul
+
+localized "notation r ` r•[` f `] ` :=
+  @@has_smul.smul (restrict_scalars.has_smul f _) r" in change_of_rings
+
+end unbundled
+
+open_locale change_of_rings
+
+variable (M : Module.{v} S)
+
+/-- Any `S`-module M is also an `R`-module via a ring homomorphism `f : R ⟶ S` by defining
+    `r • m := f r • m` (`module.comp_hom`). This is called restriction of scalars. -/
+def obj' : Module R :=
+{ carrier := M,
+  is_add_comm_group := infer_instance,
+  is_module := module.comp_hom M f }
+
+
+@[simp] lemma smul_def (r : R) (m : M) : r r•[f] m = f r • m := rfl
```

Comment:
I guess we also need this in the unbundled case?

Reply:
done